Resolve ambiguous build system with preferred choice

dpsutton opened the pull request:
Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings (4 preexisting)
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [x] You've updated the readme (if adding/changing user-visible functionality)
- [x] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

When there are artifacts from multiple build systems, eg: lein, boot,
etc, look in `cider-preferred-build-system`.

Allow setting a variable to resolve conflicts when multiple build systems are found. Ensure valid choice, otherwise do the current behavior: ask the user.

[associated issue 1672](https://github.com/clojure-emacs/cider/issues/1672)

bbatsov commented on cider.el:
```diff
@@ -146,6 +146,18 @@ project.clj for leiningen or build.boot for boot, could be found."
+  nil
+  "Allow choosing a build system when there are many.
+When there are artifacts from multiple build systems (\"lein\", \"boot\",
+\"gradle\") the user is presented with a choice that requires input. This
```
the user is prompted to select one of them

bbatsov commented on cider.el:
```diff
@@ -146,6 +146,18 @@ project.clj for leiningen or build.boot for boot, could be found."
+  "Allow choosing a build system when there are many.
+When there are artifacts from multiple build systems (\"lein\", \"boot\",
+\"gradle\") the user is presented with a choice that requires input. This
+value will prevent the choice being displayed and used the default build
```
When non-nil this variable will suppress this behaviour and will select whatever build system is indicated by the variable if present.

bbatsov commented on cider.el:
```diff
@@ -146,6 +146,18 @@ project.clj for leiningen or build.boot for boot, could be found."
+systems to use and will never override a command when there is no
+ambiguity."
+  :type 'string
+  :group 'cider)
```
This should have a `:package-version` property.

bbatsov replied:
You should also make this a choice defcustom, so people would be able to see the valid choices in the UI.

bbatsov commented on cider.el:
```diff
@@ -146,6 +146,18 @@ project.clj for leiningen or build.boot for boot, could be found."
   :group 'cider
   :package-version '(cider . "0.9.0"))
 
+(defcustom cider-preferred-command
```
I'd call this `cider-preferred-build-tool` or something like this.

Malabarba replied:
Sorry about that. Calling it a command was my suggestion. :stuck_out_tongue_closed_eyes:

bbatsov commented on cider.el:
```diff
@@ -564,22 +576,28 @@ Use `cider-ps-running-nrepls-command' and `cider-ps-running-nrepl-path-regexp-li
   "Determine the type, either leiningen, boot or gradle, of the current project.
-If more than one project file types are present, prompt the user to choose."
+If more than one project file types are present, check for a preferred
+version in `cider-preferred-command`, otherwise prompt the user to
```
version -> command / build tool

bbatsov commented on cider.el:
```diff
@@ -564,22 +576,28 @@ Use `cider-ps-running-nrepls-command' and `cider-ps-running-nrepl-path-regexp-li
-                                  ("boot" . "build.boot")
-                                  ("gradle" . "build.gradle")))))
+                                command-choices)))
+         (too-many (> (length choices) 1))
```
too-many -> multiple-options/project-files/something like this

bbatsov commented on cider.el:
```diff
@@ -564,22 +576,28 @@ Use `cider-ps-running-nrepls-command' and `cider-ps-running-nrepl-path-regexp-li
-                             choices nil t nil nil default)
-          (car choices))
-        cider-default-repl-command)))
+    (cond ((and too-many choices (assoc cider-preferred-command command-choices))
```
choices seems redundant here

bbatsov commented on cider.el:
```diff
@@ -564,22 +576,28 @@ Use `cider-ps-running-nrepls-command' and `cider-ps-running-nrepl-path-regexp-li
-        cider-default-repl-command)))
+    (cond ((and too-many choices (assoc cider-preferred-command command-choices))
+           cider-preferred-command)
+          ((and too-many choices)
```
Here as well.

bbatsov commented:
As this behaviour is getting a bit complex you should also add a few unit tests to cover it.

dpsutton commented:
@bbatsov what package version should i put? 13 or 12.1?

bbatsov commented:
0.13

dpsutton commented:
I added the tests and renamed it as you liked. I am banging my head against the wall about how to get defcustom to provide a list of choices if anyone can help out with that or provide some guidance. The documentation is not the greatest.

Malabarba commented:
(Ignore the accidental close. Fat fingers.)
You can do `: type '(choice "lein" "boot" "gradle")`.

dpsutton commented:
@Malabarba when i do that, i get the following:
![image](https://cloud.githubusercontent.com/assets/6377293/14742670/08631d10-0863-11e6-801c-ef1787453227.png)


    :type '(choice "lein" "boot" "gradle")
    :group 'cider
    :package-version '(cider . "0.13.0"))

dpsutton commented:
that could be an artifact of my build. I'm on some random commit of the 25 branch and its been a while since i've rebuilt.

Malabarba commented:
Yeah, looks like it's unrelated.

dpsutton commented:
hmm. that looks like a legitimate failure. I'll look at it at lunch

bbatsov commented on cider.el:
```diff
@@ -146,6 +146,19 @@ project.clj for leiningen or build.boot for boot, could be found."
+  nil
+  "Allow choosing a build system when there are many.
+When there are artifacts from multiple build systems (\"lein\", \"boot\",
+\"gradle\") the user is prompted to select one of them. When non-nil, this
```
You should use two spaces after `.` in docstrings. Elisp conventions...

dpsutton replied:
yeah i remember editing my first elisp file and the warnings popping up. It drove me mad haha

bbatsov commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### New Features
 
+* Allow setting a preffered command when multiple are found in `cider-preferred-command`.
```
You forgot to update this.

bbatsov commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### New Features
 
+* Allow setting a preffered build tool when multiple are found in `cider-preferred-command`.
```
preferred-build-tool

bbatsov commented on cider.el:
```diff
@@ -562,24 +575,36 @@ Use `cider-ps-running-nrepls-command' and `cider-ps-running-nrepl-path-regexp-li
           (setq paths (cons (match-string 1) paths)))))
     (seq-uniq paths)))
 
+(defun cider--identify-buildtools-present ()
```
build tools or build files?

dpsutton replied:
I'm extrapolating the presence of a build tool from the build file. I return not a list of files present but a list of build tools, ("lein" "boot" ...). So from my point of view it stays. But I can see the opposite side as well, and your opinion coming from not being in the code is probably more meaningful as you haven't internalized everything like i have.

bbatsov replied:
My point was mostly that you can't be certain if the build tools are present or not until you check for their executables. But I see your point. We can enhance the code to check if something is present or not, but this would complicate it further for a pretty marginal gain.

bbatsov commented on cider.el:
```diff
@@ -562,24 +575,36 @@ Use `cider-ps-running-nrepls-command' and `cider-ps-running-nrepl-path-regexp-li
     (seq-uniq paths)))
 
+(defun cider--identify-buildtools-present ()
+  "Try to go to all project build files in order to determine possible
```
The first line of every docstring should be a complete sentence.

bbatsov commented on doc/miscellaneous_features.md:
```diff
@@ -41,8 +41,8 @@ buffer (which is internally using `cider-macroexpansion-mode`):
 --------------------------------|-------------------------------
-<kbd>C-c C-m</kbd>              | Invoke `macroexpand-1` on the form at point and replace the original form with its expansion.  If invoked with a prefix argument, `macroexpand` is used instead of `macroexpand-1`.
-<kbd>C-c M-m</kbd>              | Invoke `clojure.walk/macroexpand-all` on the form at point and replace the original form with its expansion.
+<kbd>m</kbd>              | Invoke `macroexpand-1` on the form at point and replace the original form with its expansion.  If invoked with a prefix argument, `macroexpand` is used instead of `macroexpand-1`.
```
This commit doesn't belong here.

bbatsov commented:
The PR is good enough at this point. Fix the changelog and squash your commits together.

dpsutton commented:
Everything is squashed down. I wasn't too sure what you meant by fix the changelog though, @bbatsov .

bbatsov commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### New Features
 
+* Allow setting a preffered-build-tool when multiple are found in `cider-preferred-command`.
```
via `cider-preferred-build-tool`
